examples/examples/jsonrpsee_as_service.rs

```diff
@@ -0,0 +1,201 @@
+			// and used because it's just a toy example.
+
+			self.inner.call(req, label).await
+		} else {
+			self.inner.call(req, label).await
+		}
+	}
+}
+
+#[rpc(server, client)]
+pub trait Rpc {
+	#[method(name = "trusted_call")]
+	async fn trusted_call(&self) -> Result<String, ErrorObjectOwned>;
+}
+
+#[async_trait]
+impl RpcServer for () {
+	async fn trusted_call(&self) -> Result<String, ErrorObjectOwned> {
+		Ok("mysecret".to_string())
+	}
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+	let filter = tracing_subscriber::EnvFilter::try_from_default_env()?;
+	tracing_subscriber::FmtSubscriber::builder().with_env_filter(filter).finish().try_init()?;
+
+	let handle = run_server();
+	tokio::spawn(handle.stopped());
+
+	{
+		let client = HttpClientBuilder::default().build("http://127.0.0.1:9944").unwrap();
+
+		// Fails because the authorization header is missing.
+		let x = client.trusted_call().await.unwrap_err();
+		tracing::info!("response: {x}");
+	}
+
+	{
+		let mut headers = HeaderMap::new();
+		headers.insert(AUTHORIZATION, HeaderValue::from_static("don't care in this example"));
+
+		let client = HttpClientBuilder::default().set_headers(headers).build("http://127.0.0.1:9944").unwrap();
+
+		let x = client.trusted_call().await.unwrap();
+		tracing::info!("response: {x}");
+	}
+
+	Ok(())
+}
+
+fn run_server() -> ServerHandle {
+	use hyper::service::{make_service_fn, service_fn};
+
+	let addr = SocketAddr::from(([127, 0, 0, 1], 9944));
+	let (tx, rx) = tokio::sync::watch::channel(());
+
+	let server_handle = ServerHandle::new(tx);
+	let stop_handle = StopHandle::new(rx);
+	let svc_builder = jsonrpsee::server::Server::builder()
+		.set_http_middleware(tower::ServiceBuilder::new().layer(CorsLayer::permissive()))
+		.max_connections(33)
+		.to_service_builder();
+
+	let methods = ().into_rpc();
+	let stop_handle2 = stop_handle.clone();
+	let metrics = Metrics::default();
+
+	// And a MakeService to handle each connection...
+	let make_service = make_service_fn(move |_conn: &AddrStream| {
+		// You may use `conn` or the actual HTTP request to get connection related details.
+
+		let stop_handle = stop_handle2.clone();
+		let svc_builder = svc_builder.clone();
+		let methods = methods.clone();
+		let metrics = metrics.clone();
+
+		async move {
+			let stop_handle = stop_handle.clone();
+			let svc_builder = svc_builder.clone();
+			let methods = methods.clone();
+			let metrics = metrics.clone();
+
+			Ok::<_, Box<dyn StdError + Send + Sync>>(service_fn(move |req| {
+				let metrics = metrics.clone();
+				let svc_builder = svc_builder.clone();
+				let methods = methods.clone();
+				let stop_handle = stop_handle.clone();
+
+				async move {
+					let headers = req.headers().clone();
+
+					// NOTE, the rpc middleware must be initialized here to be able to created once per connection
+					// with data from the connection such as the headers in this example
+					let rpc_middleware = RpcServiceBuilder::new()
```

Comment:
I don't like this but that's the trade-off right now to get state to the middleware